In [18]:
library(tidyverse)
library(repr)
library(tidymodels)
options(repr.matrix.max.rows = 6)

set.seed(8205)

 # Reading in Heart Disease Data

In [19]:

#download.file("https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.switzerland.data",
            #  destfile = "./switzerland_data.data")

#read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/switzerland.data/processed.switzerland.data")

cleveland_data <- read_csv("data/processed.cleveland.data", col_names = FALSE) |>
    rename(age      = X1,      # Age in years
           sex      = X2,      # Sex (1 = male, 0 = female)
           cp       = X3,      # Chest pain type 
                               # - 1 : typical angina
                               # - 2 : atypical angina
                               # - 3 : non-anginal pain
                               # - 4 : asymptomatic
           trestbps = X4,      # Resting blood pressure (mmHg)
           chol     = X5,      # Serum cholestoral (mg/dL)
           fbs      = X6,      # Fasting blood sugar > 120 mg/dL (1 = true, 0 = false)
           restecg  = X7,      # Resting electrocardiographic results 
                               # - 0 : normal,
                               # - 1 : having ST-T wave abnormality (T wave inversions and/or ST 
                               #       elevation or depression of > 0.05 mV)
                               # - 2 : showing probable or definite left ventricular hypertrophy by Estes' criteria
           thalach  = X8,      # Maximum heart rate achieved
           exang    = X9,      # Exercise induced angina (0 = no, 1 = yes)
           oldpeak  = X10,     # ST depression induced by exercise relative to rest
           slope    = X11,     # The slope of the peak exercise ST segment
                               # - 1 : upsloping
                               # - 2 : flat
                               # - 3 : downsloping
           ca       = X12,     # Number of major vessels (0-3) colored by flourosopy
           thal     = X13,     # 3 = normal; 6 = fixed defect; 7 = reversable defect
           num      = X14)     # diagnosis of heart disease (angiographic disease status)

cleveland_data

Rows: 303 Columns: 14
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): X12, X13
dbl (12): X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X14

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca  thal
1   63  1   1  145      233  1   2       150     0     2.3     3     0.0 6.0 
2   67  1   4  160      286  0   2       108     1     1.5     2     3.0 3.0 
3   67  1   4  120      229  0   2       129     1     2.6     2     2.0 7.0 
⋮   ⋮   ⋮   ⋮  ⋮        ⋮    ⋮   ⋮       ⋮       ⋮     ⋮       ⋮     ⋮   ⋮   
301 57  1   4  130      131  0   0       115     1     1.2     2     1.0 7.0 
302 57  0   2  130      236  0   2       174     0     0.0     2     1.0 3.0 
303 38  1   3  138      175  0   0       173     0     0.0     1     ?   3.0 
    num
1   0  
2   2  
3   1  
⋮   ⋮  
301 3  
302 1  
303 0

# Data wrangling

In [20]:
cleveland_data <- cleveland_data |>
    mutate(age      = as.integer(age),
           sex      = as.factor(sex),
           cp       = as.factor(cp),
           trestbps = as.integer(trestbps),
           chol     = as.integer(chol),
           fbs      = as.factor(fbs),
           restecg  = as.factor(restecg),
           thalach  = as.integer(thalach),
           exang    = as.factor(exang),
           oldpeak  = as.double(oldpeak),
           slope    = as.factor(slope),
           ca       = as.integer(ca),
           thal     = as.factor(thal),
           num      = as.factor(num))

cleveland_data

Warning message in mask$eval_all_mutate(quo):
“NAs introduced by coercion”


age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal
1   63  1   1  145      233  1   2       150     0     2.3     3     0  6.0 
2   67  1   4  160      286  0   2       108     1     1.5     2     3  3.0 
3   67  1   4  120      229  0   2       129     1     2.6     2     2  7.0 
⋮   ⋮   ⋮   ⋮  ⋮        ⋮    ⋮   ⋮       ⋮       ⋮     ⋮       ⋮     ⋮  ⋮   
301 57  1   4  130      131  0   0       115     1     1.2     2      1 7.0 
302 57  0   2  130      236  0   2       174     0     0.0     2      1 3.0 
303 38  1   3  138      175  0   0       173     0     0.0     1     NA 3.0 
    num
1   0  
2   2  
3   1  
⋮   ⋮  
301 3  
302 1  
303 0

In [21]:
# Splitting the data into training and testing sets
cleveland_split <- initial_split(cleveland_data, prop = 0.75, strata = num)
cleveland_training <- training(cleveland_split) 
cleveland_testing <- testing(cleveland_split)

# Exploratory data analysis

### Heart Disease Summary

In [22]:
target_summary <- cleveland_training |>
    group_by(num) |>
    summarize(count = n())
target_summary

num count
1 0   123  
2 1    42  
3 2    26  
4 3    27  
5 4     8

From the dataset, the majority of patients do not have any presence of heart disease. For those who have presence of a heart disease, most have Type 1 whereas Type 4 is the least common.

### Demographics

In [23]:
demographic_summary <- cleveland_training |>
    group_by(sex) |>
    summarize(count = n(), average_age = mean(age))
demographic_summary

sex count average_age
1 0    73   56.06849   
2 1   153   53.66667

The number of male double that of female participants. The average age for this dataset is around 54 years old. In addition, the average age of the male is slightly younger than female participants.

### Blood Pressure

In [25]:
average_rest_bps <- cleveland_training |>
    summarize(average_restbps = mean(trestbps))
average_rest_bps

average_bps
1 132.3319

The average resting blood pressure of all participants is 132*mmHg*. Let's see how this would affect the type of heart disease.

In [26]:
average_restbps_by_type <- cleveland_training |>
    group_by(num) |>
    summarize(average_restbps = mean(trestbps))
average_restbps_by_type

num average_bps
1 0   128.8293   
2 1   135.7143   
3 2   136.2692   
4 3   136.5926   
5 4   141.2500

People with no presence of heart disease have lower average resting blood pressure. Those who do thend to have increasingly higher resting blood pressure.